pipeline/06-upload.R

```diff
@@ -0,0 +1,320 @@
+    paths$output$report$s3
+  )
+}
+
+
+
+
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# 3. Crawl and notify ----------------------------------------------------------
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+# This will run a Glue crawler to update schemas and send an email to any SNS
+# subscribers. Only run when actually uploading
+if (params$toggle$upload_to_s3) {
+  message("Sending run email and running model crawler")
+
+  # If assessments and SHAP values were uploaded, trigger a Glue crawler to find
+  # any new partitions
+  if (run_type == "full") {
+    glue_srv <- paws.analytics::glue()
+    glue_srv$start_crawler("ccao-model-results-crawler")
+  }
+
+  # If SNS ARN is available, notify subscribers via email upon run completion
+  if (!is.na(Sys.getenv("AWS_SNS_ARN_MODEL_STATUS", unset = NA))) {
+    pipeline_sns <- paws.application.integration::sns(
+      config = list(region = Sys.getenv("AWS_REGION"))
+    )
+
+    # Get pipeline total run time from file
+    pipeline_sns_total_time <- read_parquet(paths$output$timing$local) %>%
+      mutate(dur = lubridate::seconds_to_period(round(overall_sec_elapsed))) %>%
+      dplyr::pull(dur)
+
+    # Get overall stats by township for the triad of interest, collapsed into
+    # a plaintext table
+    pipeline_sns_results <- arrow::read_parquet(
+      paths$output$performance_test$local,
+      col_select = c("geography_type", "geography_id", "by_class", "cod")
+    ) %>%
+      filter(
+        tolower(town_get_triad(geography_id, name = TRUE)) ==
+          params$assessment$triad,
+        !by_class, geography_type == "township_code"
+      ) %>%
+      mutate(township_name = town_convert(geography_id)) %>%
+      select(cod, township_name) %>%
+      mutate(across(where(is.numeric), round, 2)) %>%
+      arrange(cod) %>%
+      knitr::kable(format = "rst") %>%
+      as.character() %>%
+      .[!grepl("=", .)] %>%
+      paste0(collapse = "\n")
+
+    # Get a link to the uploaded Quarto report
+    report_path_parts <- strsplit(paths$output$report$s3[1], "/")[[1]]
+    report_bucket <- report_path_parts[3]
+    report_path <- report_path_parts[4:length(report_path_parts)] %>%
+      paste(collapse = "/")
+    # Use direct link to the console instead of to the object so that we don't
+    # have to bother with signed URLs
+    report_url <- paste0(
+      "https://s3.console.aws.amazon.com/s3/object/",
+      "{report_bucket}/{report_path}?region=us-east-1&tab=overview"
+    ) %>%
+      glue::glue()
+
+    # Publish to SNS
+    pipeline_sns$publish(
+      Subject = paste("Model Run Complete:", run_id),
+      Message = paste0(
+        "Model run: ", run_id, " complete\n",
+        "Finished in: ", pipeline_sns_total_time, "\n\n",
+        "Report link: ", report_url, "\n\n",
+        pipeline_sns_results
+      ),
+      TopicArn = Sys.getenv("AWS_SNS_ARN_MODEL_STATUS")
+    )
+  }
```

Comment:
Adding a report link to the SNS notification is new as of this PR. Note that this hasn't been tested yet because SNS notifications won't be sent until #63 is fixed by #64.